File: android/src/com/marpies/ane/facebook/functions/SendOpenGraphRequestFunction.java
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.facebook.functions;

import android.os.Bundle;
import com.adobe.fre.FREArray;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.facebook.AccessToken;
import com.facebook.GraphRequest;
import com.facebook.GraphResponse;
import com.facebook.HttpMethod;
import com.marpies.ane.facebook.data.AIRFacebookEvent;
import com.marpies.ane.facebook.utils.AIR;
import com.marpies.ane.facebook.utils.FREObjectUtils;
import com.marpies.ane.facebook.utils.StringUtils;
import org.json.JSONObject;

public class SendOpenGraphRequestFunction extends BaseFunction {

	private static GraphRequest.Callback mGraphRequestCallback;

	/**
	 *
	 *
	 * Public API
	 *
	 *
	 */

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		AccessToken accessToken = AccessToken.getCurrentAccessToken();

		HttpMethod method = HttpMethod.valueOf( FREObjectUtils.getStringFromFREObject( args[0] ) );
		String graphPath = FREObjectUtils.getStringFromFREObject( args[1] );
		mListenerID = FREObjectUtils.getIntFromFREObject( args[3] );

		GraphRequest request = null;
		switch( method ) {
			case GET:
				request = GraphRequest.newGraphPathRequest( accessToken, graphPath, getCallback() );
				AIR.log( "Sending Open Graph GET request with path " + graphPath + "." );
				if( args.length > 2 && args[2] != null ) {
					Bundle parameters = FREObjectUtils.getBundleFromFREArray( (FREArray) args[2] );
					request.setParameters( parameters );
				}
				break;
			case POST:
				JSONObject object = null;
				if( args.length > 2 && args[2] != null ) {
					object = FREObjectUtils.getJSONObjectFromFREArray( (FREArray) args[2] );
				}
				request = GraphRequest.newPostRequest( accessToken, graphPath, object, getCallback() );
				String logMessage = "Sending Open Graph POST request with path " + graphPath;
				if( object != null ) {
					logMessage += " and parameters " + object.toString();
				}
				AIR.log( logMessage );
				break;
			case DELETE:
				request = GraphRequest.newDeleteObjectRequest( accessToken, graphPath, getCallback() );
				AIR.log( "Sending Open Graph DELETE request with ID " + graphPath + "." );
				break;
		}
		if( request != null ) {
			request.executeAsync();
		} else {
			AIR.dispatchEvent( AIRFacebookEvent.OPEN_GRAPH_ERROR,
					StringUtils.getEventErrorJSON( mListenerID, "Error creating Open Graph request." )
			);
		}

		return null;
	}

	/**
	 *
	 *
	 * Private API
	 *
	 *
	 */

	private void onGraphRequestResult( GraphResponse response ) {
		if( response.getError() != null ) {
			AIR.log( "Open Graph request error: " + response.getError().getErrorMessage() );
			AIR.dispatchEvent( AIRFacebookEvent.OPEN_GRAPH_ERROR,
					StringUtils.getEventErrorJSON( mListenerID, response.getError().getErrorMessage() )
			);
			return;
		}
		if( AIR.getContext() != null ) {
			String jsonResult = "{}";
			if( response.getJSONObject() != null ) {
				jsonResult = response.getJSONObject().toString();
			} else if( response.getJSONArray() != null ) {
				jsonResult = response.getJSONArray().toString();
			}
			AIR.dispatchEvent(
					AIRFacebookEvent.OPEN_GRAPH_SUCCESS,
					String.format( StringUtils.locale, "%d||%s||%s", mListenerID, jsonResult, response.getRawResponse() )
			);
		} else {
			AIR.log( "ANE CONTEXT IS NULL WHEN REQUESTING OPEN GRAPH" );
		}
	}

	private GraphRequest.Callback getCallback() {
		if( mGraphRequestCallback == null ) {
			mGraphRequestCallback = new GraphRequest.Callback() {
				@Override
				public void onCompleted( GraphResponse response ) {
					onGraphRequestResult( response );
				}
			};
		}
		return mGraphRequestCallback;
	}

}

File: ios/AIRFacebook/Functions/SendOpenGraphRequestFunction.m
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
#import "SendOpenGraphRequestFunction.h"
#import <Foundation/Foundation.h>
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import "AIRFacebook.h"
#import "AIRFacebookEvent.h"
#import <AIRExtHelpers/MPStringUtils.h>
#import <FBSDKCoreKit/FBSDKCoreKit.h>

FREObject fb_sendOpenGraphRequest( FREContext context, void *functionData, uint32_t argc, FREObject *argv ) {
    [AIRFacebook log:@"sendOpenGraphRequest"];

    NSString* method = [MPFREObjectUtils getNSString:argv[0]];
    NSString* path = [MPFREObjectUtils getNSString:argv[1]];
    NSDictionary* params = nil;
    /* Additional parameters are only applicable to GET and POST requests */
    if( ![method isEqualToString:@"DELETE"] && argc > 2 ) {
        params = (argv[2] == nil) ? nil : [MPFREObjectUtils getNSDictionary:argv[2]];
    }
    /* Listener ID */
    int listenerID = [MPFREObjectUtils getInt:argv[3]];

    NSMutableString* logMessage = [NSMutableString stringWithFormat:@"Sending open graph request '%@' w/ method '%@'", path, method];
    if( params ) {
        [logMessage appendFormat:@" and parameters %@", params];
    }
    [AIRFacebook log:logMessage];

    FBSDKGraphRequest* request = [[FBSDKGraphRequest alloc] initWithGraphPath:path
                                                                   parameters:params
                                                                   HTTPMethod:method];
    [request startWithCompletionHandler:^( FBSDKGraphRequestConnection* connection, id result, NSError* error ) {
        /* Request error */
        if( error ) {
            [AIRFacebook dispatchEvent:OPEN_GRAPH_ERROR
                           withMessage:[MPStringUtils getEventErrorJSONString:listenerID errorMessage:error.localizedDescription]];
        }
        /* Request success */
        else {
            /* Try to parse JSON from the result NSArray/NSDictionary */
            NSError* parseError = nil;
            NSData* jsonData = [NSJSONSerialization dataWithJSONObject:result options:0 error:&parseError];
            /* Error parsing */
            if( !jsonData ) {
                [AIRFacebook log:[NSString stringWithFormat:@"Open Graph request error: %@", parseError.localizedDescription]];
                [AIRFacebook dispatchEvent:OPEN_GRAPH_ERROR
                               withMessage:[MPStringUtils getEventErrorJSONString:listenerID errorMessage:parseError.localizedDescription]];
            }
            /* Successfully parsed */
            else {
                NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
                [AIRFacebook log:@"Open Graph request success"];
                [AIRFacebook dispatchEvent:OPEN_GRAPH_SUCCESS
                               withMessage:[NSString stringWithFormat:@"%d||%@||%@", listenerID, jsonString, [NSString stringWithFormat:@"%@", result]]];
            }
        }
    }];

    return nil;
}